server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SearchAction.java

```diff
@@ -43,47 +68,106 @@ public void define(WebService.NewController controller) {
+      QGateWithOrgDto qualityGate = qualityGateFinder.getByOrganizationAndId(dbSession, organization, request.mandatoryParamAsLong(PARAM_GATE_ID));
+      Association associations = find(dbSession,
+        ProjectQgateAssociationQuery.builder()
+          .qualityGate(qualityGate)
+          .membership(request.param(PARAM_QUERY) == null ? request.param(SELECTED) : ANY)
+          .projectSearch(request.param(PARAM_QUERY))
+          .pageIndex(request.paramAsInt(PARAM_PAGE))
+          .pageSize(request.paramAsInt(PARAM_PAGE_SIZE))
+          .build());
+
+      Qualitygates.SearchResponse.Builder createResponse = Qualitygates.SearchResponse.newBuilder()
+        .setMore(associations.hasMoreResults());
+
       for (ProjectQgateAssociation project : associations.projects()) {
-        writer.beginObject().prop("id", project.id()).prop("name", project.name()).prop(Param.SELECTED, project.isMember()).endObject();
+        createResponse.addResultsBuilder()
+          .setId(project.id())
+          .setName(project.name())
+          .setSelected(project.isMember());
       }
-      writer.endArray().endObject().close();
+
+      writeProtobuf(createResponse.build(), request, response);
     }
   }
 
+  private SearchAction.Association find(DbSession dbSession, ProjectQgateAssociationQuery query) {
+    List<ProjectQgateAssociationDto> projects = dbClient.projectQgateAssociationDao().selectProjects(dbSession, query);
+    List<ProjectQgateAssociationDto> authorizedProjects = keepAuthorizedProjects(dbSession, projects);
+
+    Paging paging = forPageIndex(query.pageIndex())
+      .withPageSize(query.pageSize())
+      .andTotal(authorizedProjects.size());
+    return new SearchAction.Association(toProjectAssociations(getPaginatedProjects(authorizedProjects, paging)), paging.hasNextPage());
+  }
+
+  private static List<ProjectQgateAssociationDto> getPaginatedProjects(List<ProjectQgateAssociationDto> projects, Paging paging) {
+    return projects.stream().skip(paging.offset()).limit(paging.pageSize()).collect(MoreCollectors.toList());
+  }
+
+  private static List<ProjectQgateAssociation> toProjectAssociations(List<ProjectQgateAssociationDto> dtos) {
+    return dtos.stream().map(ProjectQgateAssociationDto::toQgateAssociation).collect(MoreCollectors.toList());
+  }
+
+  private List<ProjectQgateAssociationDto> keepAuthorizedProjects(DbSession dbSession, List<ProjectQgateAssociationDto> projects) {
+    if (userSession.isRoot()) {
+      // the method AuthorizationDao#keepAuthorizedProjectIds() should be replaced by
+      // a call to UserSession, which would transparently support roots.
+      // Meanwhile root is explicitly handled.
+      return projects;
+    }
+    List<Long> projectIds = projects.stream().map(ProjectQgateAssociationDto::getId).collect(MoreCollectors.toList());
+    Collection<Long> authorizedProjectIds = dbClient.authorizationDao().keepAuthorizedProjectIds(dbSession, projectIds, userSession.getUserId(), UserRole.USER);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "keepAuthorizedProjectIds"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)